File: example-servers/python/flask/src/services/custom.py
from flask import Response
import time
import json

class Custom:
    def chat(self, body):
        # Text messages are stored inside request body using the Deep Chat JSON format:
        # https://deepchat.dev/docs/connect
        print(body)
        # Sends response back to Deep Chat using the Result format:
        # https://deepchat.dev/docs/connect/#Result
        return {'result': {'text': 'This is a respone from a Flask server. Thankyou for your message!'}}

    def chat_stream(self, body):
        # Text messages are stored inside request body using the Deep Chat JSON format:
        # https://deepchat.dev/docs/connect
        print(body)
        response_chunks = 'This is a response from a Flask server. Thank you for your message!'.split(
            ' ')
        response = Response(self.send_stream(response_chunks), mimetype='text/event-stream')
        response.headers['Content-Type'] = 'text/event-stream'
        response.headers['Cache-Control'] = 'no-cache'
        response.headers['Connection'] = 'keep-alive'
        response.headers['Access-Control-Allow-Origin'] = '*'
        return response

    def send_stream(self, response_chunks, chunk_index=0):
        if chunk_index < len(response_chunks):
            chunk = response_chunks[chunk_index]
            yield f"data: {json.dumps({'result': {'text': f'{chunk} '}})}\n\n"
            time.sleep(0.07)
            yield from self.send_stream(response_chunks, chunk_index + 1)
        else:
            yield ''

    def files(self, request):
        # Files are stored inside a files object
        files = request.files.getlist('files')
        if files:
            print('Files:')
            for file in files:
                print(file.filename)

            # When sending text messages along with files - they are stored inside the data form
            textMessages = list(request.form.items())
            if len(textMessages) > 0:
                print('Text messages:')
                for key, value in textMessages:
                    print(key, value)
        else:
            # When sending text messages without any files - they are stored inside a json
            print("Text messages:")
            print(request.json)

        # Sends response back to Deep Chat using the Result format:
        # https://deepchat.dev/docs/connect/#Result
        return {'result': {'text': 'This is a respone from a Flask server. Thankyou for your message!'}}

File: example-servers/node/nestjs/src/services/custom.ts
import {Injectable} from '@nestjs/common';
import {Request, Response} from 'express';

@Injectable()
export class Custom {
  async chat(body: Request['body']) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    console.log(body);
    // Sends response back to Deep Chat using the Response format:
    // https://deepchat.dev/docs/connect/#Response
    return {text: 'This is a respone from a NestJs server. Thankyou for your message!'};
  }

  async chatStream(body: Request['body'], res: Response) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    console.log(body);
    const responseChunks = 'This is a respone from a NestJs server. Thankyou for your message!'.split(' ');
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Connection', 'keep-alive');
    res.setHeader('Access-Control-Allow-Origin', '*');
    Custom.sendStream(res, responseChunks);
  }

  private static sendStream(res: Response, responseChunks: string[], chunkIndex = 0) {
    setTimeout(() => {
      const chunk = responseChunks[chunkIndex];
      if (chunk) {
        // Sends response back to Deep Chat using the Response format:
        // https://deepchat.dev/docs/connect/#Response
        res.write(`data: ${JSON.stringify({text: `${chunk} `})}\n\n`);
        Custom.sendStream(res, responseChunks, chunkIndex + 1);
      } else {
        res.end();
      }
    }, 70);
  }

  async files(files: Express.Multer.File[], body: Request['body']) {
    // Files are stored inside a form using Deep Chat request FormData format:
    // https://deepchat.dev/docs/connect
    if (files) {
      console.log('Files:');
      console.log(files);
    }
    // When sending text along with files, it is stored inside the request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    if (Object.keys(body).length > 0) {
      console.log('Text messages:');
      // message objects are stored as strings and they will need to be parsed (JSON.parse) before processing
      console.log(body);
    }
    // Sends response back to Deep Chat using the Response format:
    // https://deepchat.dev/docs/connect/#Response
    return {text: 'This is a respone from a NestJs server. Thankyou for your message!'};
  }
}
